### ISSUE TITLE
JavaSource Language is very inperformant

### ISSUE THREAD

**Issue by cdietrich** [eclipse/xtext-extras#261]
the JavaSource Language is very inperformant e.g. when used in the context of the xtend gradle plugin (https://github.com/xtext/xtext-gradle-plugin/issues/89).

The main performance downside in the `installFull` in  `JavaDerivedStateComputer` which makes inefficient use of a own `IndexAwareNameEnvironment` and the jdt compiler

here is a small but "unfair" sample that shows the problem quite nicely i think

```

import com.google.inject.Inject
import com.google.inject.Provider
import java.util.List
import org.eclipse.emf.common.util.URI
import org.eclipse.xtext.common.types.access.IJvmTypeProvider
import org.eclipse.xtext.java.JavaSourceLanguageSetup
import org.eclipse.xtext.resource.IResourceDescription
import org.eclipse.xtext.resource.XtextResourceSet
import org.eclipse.xtext.resource.impl.ChunkedResourceDescriptions
import org.eclipse.xtext.resource.impl.ResourceDescriptionsData
import org.eclipse.xtext.testing.util.InMemoryURIHandler

class DummyTest {
	
	def static  void main(String[] args) {
		val injector = new JavaSourceLanguageSetup().createInjectorAndDoEMFRegistration
		val dt = injector.getInstance(DummyTest)
		dt.doit()
	}
	
	    @Inject Provider<XtextResourceSet> resourceSetProvider
    @Inject IResourceDescription.Manager resourceDesriptionManager
    @Inject IJvmTypeProvider.Factory typeProviderFactory
	
	def void doit() {
		val max = 10000
		val rs = (1..max).map["demo/Demo"+it+".java"->'''package demo; public class Demo«it» {
			private String x;
			«IF it > 1»
			private Demo«it-1» demo;
			«ENDIF»
			«IF it > 2»
			private Demo«it-2» demo;
			«ENDIF»
			«IF it < max-1»
			private Demo«it+1» demo;
			«ENDIF»
		}'''].toList.resourceSet
		println("creation done")
		for (r : rs.resources) {
			r.contents
		}
	}
	
	    def protected resourceSet(Pair<String, String> ... files) {
        val result = resourceSetProvider.get
        typeProviderFactory.createTypeProvider(result)
        result.classpathURIContext = class.classLoader
        result.URIConverter.URIHandlers.clear
        val uriHandler = new InMemoryURIHandler
        val uris = files.map [
            val uri = URI.createURI(InMemoryURIHandler.SCHEME + ":/" + key)
            val out = uriHandler.createOutputStream(uri, emptyMap)
            out.write(value.bytes)
            out.close
            return uri
        ]
        result.URIConverter.URIHandlers.add(uriHandler)
        val List<IResourceDescription> descriptions = newArrayList()
        for (uri : uris) {
            val resource = result.getResource(uri, true)
            descriptions += resourceDesriptionManager.getResourceDescription(resource);
        }
        val chunkedResourceDescriptions = new ChunkedResourceDescriptions(#{
            'default' -> new ResourceDescriptionsData(descriptions)
        })
        (chunkedResourceDescriptions).attachToEmfObject(result)
        return result
    }
	
}
```

**Comment by cdietrich** [comment]
i wonder if

- we can reuse the name environment
- the compiler
- can "installFull" other resources as side effect of compiling the current resource
- can set different compiler options

**Comment by cdietrich** [comment]
with photon the problem is even worse 
https://github.com/xtext/xtext-gradle-plugin/issues/120

**Comment by cdietrich** [comment]
Fixed in Xtext 2.15
